stdlib/public/core/StringIndex.swift

```diff
@@ -524,16 +524,10 @@ extension String.Index {
     }
   }
 
-  // The definitions below are placeholders for potential future `String.Index`
-  // conformances to `CustomStringConvertible` and
-  // `CustomDebugStringConvertible`. They are supplied here to make working with
-  // string indices somewhat bearable while we're working on adding the actual
-  // conformances.
-
   /// A textual representation of this instance.
-  @_alwaysEmitIntoClient
+  @backDeployed(before: SwiftStdlib 6.1)
```

Comment:
When the `Sequence.count(where:)` method was re-implemented, there was a [comment](https://github.com/swiftlang/swift/pull/70816#discussion_r1448211636) that the standard library can't effectively use `@backDeployed` yet.

Reply:
I'm not sure what the issue was back then, but hopefully it's been resolved since then 🤞

Reply:
One seemingly minor issue that keeps causing trouble is that `@backDeployed` does not gracefully handle mid-release ABI transitions -- people whose SDK is out of phase with their running OS may end up building code that fails to link at runtime.

Another issue I recently discovered is that `@backDeployed` functions cannot be used to fulfill a member requirement in a retroactive protocol conformance. It is tempting to try backdeploying the CustomStringConvertible conformance by putting this in a public package somewhere:

```
extension String.Index: @retroactive CustomStringConvertible {
 // `description` comes straight from the stdlib
}
```

However, if `description` is marked `@backDeployed`, then this fails at runtime on versions predating its introduction. (This is clearly a bug, as `@_alwaysEmitIntoClient` functions do work in the same context.)

* * *

I think we'll want to keep `String.Index.description` as opaque as possible -- it cannot be defined fully `@_aeic`, or we would give up the flexibility to change its implementation later. At the same time, I also want the function to deploy back to earlier versions.

This is the alternative I'm considering -- it's kind of a homegrown `@backDeploy`, built from chewing gum and twine.

```swift
extension String.Index {
  @_alwaysEmitIntoClient
  public var description: String {
    if #available(SwiftStdlib 6.1, *) {
      return _descriptionImpl()
    }
    return _description
  }

  @_alwaysEmitIntoClient
  @inline(never)
  public var _description: String { // Preexisting member
    // 23[utf8]+1
    var d = "\(_encodedOffset)[\(_encodingDescription)]"
    if transcodedOffset != 0 {
      d += "+\(transcodedOffset)"
    }
    return d
  }

  @available(SwiftStdlib 6.1, *)
  public func _descriptionImpl() -> String {
    _description
  }
}
```

I don't love that the exported symbol name is different from `String.description`, but this may be a workable alternative. It makes it easy to tweak the description in future releases of the stdlib, without affecting its result when run on older versions.

Luckily, these are implementation details that do not matter for Swift Evolution. (We do promise that the property will be backdeployable, but the mechanism for that and the precise list of new symbols that the stdlib will expose are not normative.)

Reply:
> This is the alternative I'm considering -- it's kind of a homegrown `@backDeploy`, built from chewing gum and twine.

If that alternative works, perhaps it should be documented alongside [CRISP](https://github.com/swiftlang/swift/blob/main/docs/StandardLibraryProgrammersManual.md#the-curiously-recursive-inlinable-switch-pattern-crisp). However, if the pre-existing `_description` can still be used, why does `description` need to be back-deployable, when the conformance to `CustomStringConvertible` cannot be back-deployed?

Reply:
Swift does not allow the stdlib to backdeploy the conformance, but it also doesn't forbid user code to ship a retroactive conformance with the same effect:

> It is tempting to try backdeploying the CustomStringConvertible conformance by putting this in a public package somewhere:
> 
> ```swift
> extension String.Index: @retroactive CustomStringConvertible {
>  // `description` comes straight from the stdlib
> }
> ```

I am toying with the idea to suggest publishing an official stdlib companion package to ship such things. (This is not crucial for `CustomStringConvertible`, but back-deployment will be very desirable for all the upcoming `Equatable`/`Hashable` conformances on string views -- we cannot live without them.)